Implement qc pipeline for vector work

hardingnj opened the pull request:
- working on getting the qc pipeline set up.

Commit:
Updates to qc pipeline

- change invocation
- enforce configfile passed on command line
- sampleset now key component of snakefile
- added code in config for checkout.

Commit:
Fix so vector-ops repo location can be passed

- now passed via command line.

Commit:
tidy up sex call assignment

Commit:
Add backup of configuration file

- add dump of config
- report mean and mode ratio
- add min/max for male and female coverage ratios

hardingnj commented:
Ready for some feedback on this Alex, mainly "structural" rather than nitty gritty (for now at least).

Particularly the passing of the `wd` and the location of the repo, I think may be a bit unweildy.

amakunin commented on pipelines/qc/config.yml:
```diff
@@ -4,12 +4,15 @@ minimum_coverage: 10
+min_male_xratio: 0.4
+max_male_xratio: 0.6
 max_pc_contam: 4.5
+manifest: manifest
```
I could not find an example manifest. At the same time, certain format is expected.

amakunin commented on pipelines/qc/config.yml:
```diff
@@ -4,12 +4,15 @@ minimum_coverage: 10
 max_pc_contam: 4.5
+manifest: manifest
+genotyped_sites: /kwiat/vector/observatory/resources/agam/ag.allsites.nonN.zarr.zip
+allele_frequencies: /kwiat/vector/observatory/resources/agam/phase2_biallelic_variantsites_af.zarr
```
README in pipeline folder or comments within config are needed - apart from description of parameters, potential sources (or how-to-get?) of `genotyped_sites` and `allele_frequencies` can be pointed out.

hardingnj replied:
yes- more docs are needed for sure.

amakunin commented on pipelines/qc/Snakefile:
```diff
@@ -1,32 +1,69 @@
-    contam=expand("output/{sample}.contamination.csv", sample=samples.index.tolist()),
-    alignm=expand("output/{sample}.callstats.csv", sample=samples.index.tolist())
+# read_fofn
+fofn_path = os.path.join(config["vector_ops"], "tracking", sampleset, "agam.fofn.tsv")
```
tight link to vector-ops

amakunin replied:
Maybe move the entire fofn path to config?

amakunin commented on pipelines/qc/Snakefile:
```diff
@@ -35,38 +72,27 @@ rule estimate_contamination:
-    os.path.join(config["scripts_loc"], "combine_qc_cohort_stats.py")
-    
+# required for qc_report.
+rule make_manifest:
```
Alternatively, we could pass the list of samples as an argument to the `qc_report` script, thus avoiding creation of extra file. Are there downsides of this approach?

hardingnj replied:
It's slightly awkward to pass lists within snakemake I think.

amakunin replied:
I see logic in that now - thus we keep file listing samples in the output folder. This is reasonable for me.

amakunin commented on pipelines/qc/Snakefile:
```diff
@@ -1,32 +1,69 @@
+    csv=expand("{sample}.callstats.csv", sample=samples),
+    npy=expand("{sample}.covhist.npz", sample=samples),
+    contam=expand("{sample}.contamination.csv", sample=samples),
+    cfg="configuration.yml"
```
Why we cannot specify original config here?

hardingnj replied:
This forces snakemake to make a copy of `config.yml` to provide a record of the file in the output dir.

amakunin replied:
smart, agreed

amakunin commented on pipelines/qc/Snakefile:
```diff
@@ -1,32 +1,69 @@
+samples = fofn['sample'].unique()
+
+# set script dir
+script_dir = "../../scripts"
```
According to snakemake conventions (verbosity > compactness), we should either move this to config or do not use the variable at all. Given that scripts location is relative to Snakefile and we are not planning to move it for now, second option seems plausible

amakunin commented on pipelines/qc/Snakefile:
```diff
@@ -1,32 +1,69 @@
 
-samples = pd.read_csv(config["manifest"], index_col=0)
-script_dir = config["scripts_loc"] 
+sampleset = os.path.split(os.getcwd())[1]
```
What if we infer sample set from path to tracking and create the top-level directory for this set in the output `-d`?

amakunin commented:
Hi @hardingnj, great job! My major suggestion is to reverse logic in sample set inference, i.e. get it from the directory of fofn location instead of setting explicitly. The resulting command would be `bash submit.sh qc_report -p -d /path/to/output/qc/ --config fofn={path/to/tracking/sample_set/fofn} --configfile config.yml`.

Also, having a README alongside with pipeline would be cool.

amakunin commented:
Also, I like the idea of copying the config to some location within sample processing repo (vector-ops). Not sure whether to keep it in a single copy or create an instance for each sample set - in case of parameter changes (e.g., due to coverage fluctuations in WGA samples).

hardingnj commented:
> My major suggestion is to reverse logic in sample set inference, i.e. get it from the directory of fofn location instead of setting explicitly. The resulting command would be `bash submit.sh qc_report -p -d /path/to/output/qc/ --config fofn={path/to/tracking/sample_set/fofn} --configfile config.yml`.

I think this is a better way to do it- agreed. I'll make those changes.

> Also, having a README alongside with pipeline would be cool.

Yes- I should probably copy the guide (more detail needed though) to here rather than `ops`.

hardingnj commented:
The other thing I find awkward is that the fofn should be a pure input file defined by rules. 

However, as the output of this is a list (samples) that goes into `expand`, it's difficult to do this. As it stands, `snakemake` won't detect changes to the fofn and be able to update accordingly. Am I missing some snakemake technique to handle this I wonder?

hardingnj commented:
Thanks @amakunin - very helpful

amakunin commented:
> The other thing I find awkward is that the fofn should be a pure input file defined by rules.

> However, as the output of this is a list (samples) that goes into `expand`, it's difficult to do this. As it stands, `snakemake` won't detect changes to the fofn and be able to update accordingly. Am I missing some snakemake technique to handle this I wonder?

Not sure I understand that, can you expand a bit?

hardingnj commented:
So- the first thing the snakefile does is to read in the fofn. It then determines the `samples` variable- then from those it determines the files it needs to create.

If the fofn changed between runs- the `snakemake` solver wouldn't detect it- and wouldn't know it had to update the pipeline.

amakunin commented:
I see. What if you specify fofn as another input for the `qc_report` rule? In this case snakemake should account for its timestamp

hardingnj commented:
> I see. What if you specify fofn as another input for the `qc_report` rule? In this case snakemake should account for its timestamp

Actually- that would work perfectly- doh! Cheers

Commit:
Refactor snakemake pipeline

- focused around fofn not output path
- clarified some code
- added comments to config
- added readme file

NOTE: This still contains a hack for checking files exist. To be removed before production.

amakunin commented on pipelines/qc/README.md:
````diff
@@ -0,0 +1,90 @@
+bash submit.sh qc_report -p -d /kwiat/vector/observatory/production/vo_agam/qc --config fofn={path/to/fofn} --configfile config.yml
+```
+
+The path to the `fofn` must be specified. This path is used to interpolate the name of the sampleset using a regex. The sampleset is used to define the output directory.
````
Example?

amakunin commented on pipelines/qc/README.md:
```diff
@@ -0,0 +1,90 @@
+
+This pipeline creates a file in `/kwiat/vector/observatory/production/vo_agam/qc/AG1000G-UG `, `qc_summary.tsv` which should be copied to `vector-ops/tracking/AG1000G-UG/qc` (or equivalent).
+
+### Requirements
```
Also, genotyped sites and allele frequencies

amakunin commented on pipelines/qc/README.md:
```diff
@@ -0,0 +1,90 @@
+
+In `docs/guides` there is a `qc_visualisation.ipynb`. This runs through the diagnostics of an example sample set. 
+
+This forms the basis of the diagnostics for each sample set and copied to `vector-ops/tracking/{sampleset}/qc` when complete.
```
Not understanding. is copying done automatically?

amakunin commented on pipelines/qc/Snakefile:
```diff
@@ -1,36 +1,38 @@
 
 # read_fofn
-fofn_path = os.path.join(config["vector_ops"], "tracking", sampleset, "agam.fofn.tsv")
+pat = re.compile("^.+(AG1000G-\D{2}(-\D)?)/agam.fofn.tsv$")
```
This will break on VOBS sample sets. Why we cannot just use `fofn_path.split('/')[-2]`?

amakunin commented on pipelines/qc/Snakefile:
```diff
@@ -1,36 +1,38 @@
-    npy=expand("{sample}.covhist.npz", sample=samples),
-    contam=expand("{sample}.contamination.csv", sample=samples),
-    cfg="configuration.yml"
+    manifest=os.path.join(sampleset, "manifest"),
```
Purely stylistic remark: Snakemake encourages using `"{sampleset}/manifest"` in such cases. I'm not sure, but probably you can set a global wildcard by inserting 
```
wildcard_constraints:
    sampleset=sampleset
```
which should set wildcard `sampleset` constraint to string value of `sampleset`

amakunin replied:
Also applicable to further `os.path.join`

amakunin commented on pipelines/qc/Snakefile:
```diff
@@ -39,16 +41,16 @@ rule qc_report:
     max_female_xratio=config["max_female_xratio"],
     max_pc_contam=config["max_pc_contam"],
-    path=lambda y: "{sample}"
+    path=lambda y: os.path.join(sampleset, "{sample}")
```
But here the above method will not work without tweaking(

amakunin commented on pipelines/qc/config.yml:
```diff
@@ -1,18 +1,47 @@
 input_location: /kwiat/vector/mirror/observatory/vo_agam/genotype_zarrs
+
+# output location for the QC work
+output_location: /kwiat/vector/observatory/production/vo_agam/qc/{sampleset}/
```
I believe output location is not needed

Commit:
-T deprecated

Commit:
Use snakefile in this directory

Commit:
Further pipeline work

- incorporate Alex suggestions
- rework for submodule structure
- improve readme with step by step
- remove unneeded cruft from config

Commit:
rename config file for consistency

Commit:
change sampleset interpolation description

hardingnj commented:
I think this is ready for another look now @amakunin. Thanks!

Commit:
Add reqs

Commit:
fix indent

Commit:
fix indent take 2

Commit:
Adjust filepaths rel/abs

Commit:
remove logging

Commit:
Fix filepath

amakunin commented on pipelines/qc/submit.sh:
```diff
@@ -1,8 +1,6 @@
-snakemake -T $@ \
-    --cluster 'qsub -v PATH="/home/njh/miniconda3/bin:$PATH" -j y -o $(pwd)/_log/ -b n {params.pe} -l {params.req},h=!foxtrot.well.ox.ac.uk -S /bin/bash' \
+snakemake --snakefile `dirname $0`/Snakefile $@ \
+    --cluster 'qsub -v PATH="/home/njh/miniconda3/bin:$PATH" -j y -b n -l {params.req},h=!foxtrot.well.ox.ac.uk -S /bin/bash' \
```
`PATH="/home/njh/miniconda3/bin:$PATH"` seems to be quite specific. Can we use binder environment instead?

hardingnj replied:
oops well spotted!

amakunin commented on pipelines/qc/README.md:
```diff
@@ -0,0 +1,138 @@
+
+### 3. Create the configfile from the example file provided.
+
+In `vector-tools/pipelines/qc` the `config.yml` file is that used by the AG1000G project. Copy this file into an appropriate place in your repo (for `vector-ops` we used `pipelines/qc/config.yml`, and edit the filepaths and filter settings where necessary. 
```
Storing 'config.yml` in a single location per repo looks fragile. Imagine we want to adjust filter settings for a subset of samples sets.

hardingnj replied:
Yeah- I wondered about this. I'm not sure, encouraging the use of the same config feels like the sensible thing to do. I'll make a note in the README.

amakunin commented on pipelines/qc/README.md:
````diff
@@ -0,0 +1,138 @@
+  --config fofn=${pwd}/tracking/AG1000G-UG/agam.fofn.tsv \
+  --configfile vector-tools/pipelines/qc/config.yml \
+  -d /kwiat/vector/observatory/production/vo_agam/qc
+```
````
Another part of the story is that we do not leave any trace of actually running the pipeline on specific dataset in github.

hardingnj replied:
Yes- that's the intention

amakunin reviewed (commented):
Hi @hardingnj, looks good. I only have some minor comments. Looking forward to test it on funestus!

Commit:
Fix conda path

hardingnj commented:
Thanks Alex. I think we still have a few details to iron out, mainly around relative and absolute filepaths. 
The other big gap is documentation on the allele frequencies input to the contamination script.

Having said that, I'm going to merge now- running a branch that uses a submodule from a branch is too much for me to handle currently.

alimanfoo commented:
Hi Nick, just catching up on some old notifications. Are there any
outstanding comments from Alex that we should capture into new issues, so
we don't forget about them?

On Thu, 21 Feb 2019 at 15:17, Nick Harding <notifications@github.com> wrote:

> Merged #3 <https://github.com/malariagen/vector-tools/pull/3> into master.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/malariagen/vector-tools/pull/3#event-2155272260>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAq8QuIDYLFUKpACzc00mhZArtm0-sKWks5vPrilgaJpZM4anhhy>
> .
>


-- 

Alistair Miles
Head of Epidemiological Informatics
Centre for Genomics and Global Health
Big Data Institute
Li Ka Shing Centre for Health Information and Discovery
University of Oxford
Old Road Campus
Headington
Oxford
OX3 7LF
United Kingdom
Phone: +44 (0)1865 743596 or +44 (0)7866 541624
Email: alimanfoo@googlemail.com
Web: http://a <http://purl.org/net/aliman>limanfoo.github.io/
Twitter: @alimanfoo <https://twitter.com/alimanfoo>

Please feel free to resend your email and/or contact me by other means if
you need an urgent reply.

hardingnj commented:
Everything is captured in:
https://github.com/malariagen/vector-tools/issues/12
